### DISCUSSION TITLE
Why still use AAudioServiceEndpointMMAP even setting shared mode

### DISCUSSION THREAD

**Discussion by xzhan96** [google/oboe#1893] [General]
Hi @philburk 

I see the android source code here https://cs.android.com/android/platform/superproject/main/+/refs/heads/main:frameworks/av/services/oboeservice/AAudioEndpointManager.cpp;l=147;bpv=0;bpt=0

It is explicitly that if use AAUDIO_SHARING_MODE_SHARED, an AAudioServiceEndpointShared is created and used. 
However, when debugging in my phones, Xiaomi- MI 9 and Oneplus 8(IN2010), I got below logs shows AAudioServiceEndpointMMAP is used. 

> 2023-08-11 15:52:41.514 20617-20617 AudioStreamBuilder      com.mobileer.oboetester              I  rate   =      0, channels  = 0, format   = 5, **sharing = SH**, dir = INPUT
> 2023-08-11 15:52:41.514 20617-20617 AudioStreamBuilder      com.mobileer.oboetester              I  device =      0, sessionId = -1, perfMode = 12, callback: ON with frames = 0
> 2023-08-11 15:52:41.514 20617-20617 AudioStreamBuilder      com.mobileer.oboetester              I  usage  =      0, contentType = 2, inputPreset = 6, allowedCapturePolicy = 0
> 2023-08-11 15:52:41.514 20617-20617 <no-tag>                com.mobileer.oboetester              D  PlayerBase::PlayerBase()
> 2023-08-11 15:52:41.517   920-11114 <no-tag>                audioserver                          D  PlayerBase::PlayerBase()
> 2023-08-11 15:52:41.518   920-11114 APM_AudioPolicyManager  audioserver                          D  getInputForAttr() source 6, samplingRate 48000, format 0x1, channelMask 0xc, session 5593, flags 0x10 attributes={ Content type: AUDIO_CONTENT_TYPE_MUSIC Usage: AUDIO_USAGE_UNKNOWN Source: AUDIO_SOURCE_VOICE_RECOGNITION Flags: 0x100 Tags:  }
> 2023-08-11 15:52:41.520   920-11114 AudioPolic...agerCustom audioserver                          V  getPackageName: No packages for calling UID
> 2023-08-11 15:52:41.520   770-770   audio_hw_primary        android.hardware.audio@2.0-service   D  adev_open_input_stream is_karaoke_fastcapture=0
> 2023-08-11 15:52:41.520   770-770   audio_hw_primary        android.hardware.audio@2.0-service   D  adev_open_input_stream: enter: sample_rate(48000) channel_mask(0xc) devices(0x80000004)        stream_handle(0xe8682a00) io_handle(2366) source(6) format 1
> 2023-08-11 15:52:41.520   770-770   audio_hw_primary        android.hardware.audio@2.0-service   D  adev_open_input_stream is_karaoke_fastcapture=1(mmap)
> 2023-08-11 15:52:41.520   770-770   audio_hw_utils          android.hardware.audio@2.0-service   W  audio_extn_utils_update_stream_input_app_type_cfg: App type could not be selected. Falling back to default
> 2023-08-11 15:52:41.522   920-8262  AudioFlinger            audioserver                          I  AudioFlinger's thread 0x79378ec800 tid=8262 ready to run
> 2023-08-11 15:52:41.524   920-11114 APM_AudioPolicyManager  audioserver                          D  getInputForAttr() returns input 2366 type 0 selectedDeviceId 18 for port ID 1543
> 2023-08-11 15:52:41.524   920-11114 **AAudioServ...dpointMMAP** audioserver                          D  open() mMapClient.uid = 1041, pid = 920 => portHandle = 1543
> 2023-08-11 15:52:41.524   920-11114 AAudioServ...dpointMMAP audioserver                          D  open() deviceId = 18, sessionId = -1

Then I see this logic
![image](https://github.com/google/oboe/assets/13726257/964975fc-3e08-4b4d-95fe-5d462f584b89)

It seems force use AAUDIO_SHARING_MODE_EXCLUSIVE and then call mStreamInternal->open(builder) again through Binder, this maybe why use AAudioServiceEndpointMMAP finnally.

My question is why design like this? what the real purpose of AAudioServiceEndpointShared? 

Thanks a lot.

**Comment by philburk** [comment]
The EXCLUSIVE stream is used to play the output of the mixer.

There can be multiple SHARED mode streams. They all get mixed together in AAudioServiceEndpointPlay::callbackLoop()
That mixed results has to be output somehow. 

What is an easy way to output a single stream? Call AAudio and open an EXCLUSIVE stream.
All the code I needed already existed. So there was no need to write new code to pass data from the mixer to the MMAP buffer.

> call mStreamInternal->open(builder) again through Binder, 

No. When running in the service, we bypass Binder and just call the appropriate service routines directly.

**Reply by xzhan96** [reply]
Thanks a lot, understood.
